library panics when trying to marshal a nil *time.Time struct member

The default AWS dynamodb library marshals a nil pointer as "null" (the desired behavior, as would omitting the field). This library panics:

```
panic: value method time.Time.MarshalText called using nil *Time pointer

goroutine 1 [running]:
time.(*Time).MarshalText(0x0, 0x49b7300, 0x0, 0x5386740, 0x0, 0x1)
	<autogenerated>:1 +0x93
github.com/guregu/dynamo.marshal(0x49b7300, 0x0, 0x0, 0x0, 0x49b7300, 0x0, 0x0)
	/Users/bryan/go/pkg/mod/github.com/guregu/dynamo@v1.2.1/encode.go:117 +0x23e
github.com/guregu/dynamo.marshalStruct(0x499f680, 0xc0004da2d0, 0x99, 0x4b60760, 0x499f680, 0xc0004da2d0)
	/Users/bryan/go/pkg/mod/github.com/guregu/dynamo@v1.2.1/encode.go:91 +0x4ef
github.com/guregu/dynamo.marshalItem(0x499f680, 0xc0004da2d0, 0x199, 0x499f601, 0x499f680)
	/Users/bryan/go/pkg/mod/github.com/guregu/dynamo@v1.2.1/encode.go:33 +0x28a
github.com/guregu/dynamo.marshalItem(0x486d500, 0xc0004da1e0, 0x0, 0x0, 0xc00012a400)
	/Users/bryan/go/pkg/mod/github.com/guregu/dynamo@v1.2.1/encode.go:31 +0x323
github.com/guregu/dynamo.Table.Put(...)
	/Users/bryan/go/pkg/mod/github.com/guregu/dynamo@v1.2.1/put.go:24
```

--- a/encode.go
+++ b/encode.go
@@ -487,3 +487,40 @@ func isAVEqual(a, b *dynamodb.AttributeValue) bool {
 	}
 	return false
 }
+
+// isNil returns true if v is considered nil
+// this is used to determine if an attribute should be set or removed
+func isNil(v interface{}) bool {
+	if v == nil || v == "" {
+		return true
+	}
+
+	// consider v nil if it's a special encoder defined on a value type, but v is a pointer
+	rv := reflect.ValueOf(v)
+	switch v.(type) {
+	case Marshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalDynamo"); ok {
+				return true
+			}
+		}
+	case dynamodbattribute.Marshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalDynamoDBAttributeValue"); ok {
+				return true
+			}
+		}
+	case encoding.TextMarshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalText"); ok {
+				return true
+			}
+		}
+	default:
+		// e.g. (*int)(nil)
+		return rv.Kind() == reflect.Ptr && rv.IsNil()
+	}
+
+	// non-pointers or special encoders with a pointer receiver
+	return false
+}

--- a/encode.go
+++ b/encode.go
@@ -105,15 +105,34 @@ func Marshal(v interface{}) (*dynamodb.AttributeValue, error) {
 }
 
 func marshal(v interface{}, special string) (*dynamodb.AttributeValue, error) {
+	rv := reflect.ValueOf(v)
 	switch x := v.(type) {
 	case *dynamodb.AttributeValue:
 		return x, nil
 	case Marshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalDynamo"); ok {
+				// MarshalDynamo is defined on value type, but this is a nil ptr
+				return nil, nil
+			}
+		}
 		return x.MarshalDynamo()
 	case dynamodbattribute.Marshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalDynamoDBAttributeValue"); ok {
+				// MarshalDynamoDBAttributeValue is defined on value type, but this is a nil ptr
+				return nil, nil
+			}
+		}
 		av := &dynamodb.AttributeValue{}
 		return av, x.MarshalDynamoDBAttributeValue(av)
 	case encoding.TextMarshaler:
+		if rv.Kind() == reflect.Ptr && rv.IsNil() {
+			if _, ok := rv.Type().Elem().MethodByName("MarshalText"); ok {
+				// MarshalText is defined on value type, but this is a nil ptr
+				return nil, nil
+			}
+		}
 		text, err := x.MarshalText()
 		if err != nil {
 			return nil, err
@@ -125,7 +144,7 @@ func marshal(v interface{}, special string) (*dynamodb.AttributeValue, error) {
 	case nil:
 		return nil, nil
 	}
-	return marshalReflect(reflect.ValueOf(v), special)
+	return marshalReflect(rv, special)
 }
 
 var nilTm encoding.TextMarshaler

--- a/encoding_test.go
+++ b/encoding_test.go
@@ -3,9 +3,11 @@ package dynamo
 import (
 	"encoding"
 	"strconv"
+	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
 )
 
 const (
@@ -211,6 +213,10 @@ var itemEncodingTests = []struct {
 			EmptyM     map[string]bool
 			EmptyPtr   *int
 			EmptyIface interface{}
+			NilTime    *time.Time
+			NilCustom  *customMarshaler
+			NilText    *textMarshaler
+			NilAWS     *dynamodbattribute.UnixTime
 		}{
 			OK:     "OK",
 			EmptyL: []int{},

--- a/encoding_test.go
+++ b/encoding_test.go
@@ -168,6 +168,17 @@ var itemEncodingTests = []struct {
 		},
 	},
 	{
+		name: "pointer (value receiver TextMarshaler)",
+		in: &struct {
+			A *textMarshaler
+		}{
+			A: new(textMarshaler),
+		},
+		out: map[string]*dynamodb.AttributeValue{
+			"A": &dynamodb.AttributeValue{S: aws.String("false")},
+		},
+	},
+	{
 		name: "rename",
 		in: struct {
 			A string `dynamo:"renamed"`
@@ -390,9 +401,31 @@ func (tm *textMarshaler) UnmarshalText(text []byte) error {
 	return nil
 }
 
+type ptrTextMarshaler bool
+
+func (tm *ptrTextMarshaler) MarshalText() ([]byte, error) {
+	if tm == nil {
+		return []byte("null"), nil
+	}
+	if *tm {
+		return []byte("true"), nil
+	}
+	return []byte("false"), nil
+}
+
+func (tm *ptrTextMarshaler) UnmarshalText(text []byte) error {
+	if string(text) == "null" {
+		return nil
+	}
+	*tm = string(text) == "true"
+	return nil
+}
+
 var (
 	_ Marshaler                = new(customMarshaler)
 	_ Unmarshaler              = new(customMarshaler)
 	_ encoding.TextMarshaler   = new(textMarshaler)
 	_ encoding.TextUnmarshaler = new(textMarshaler)
+	_ encoding.TextMarshaler   = new(ptrTextMarshaler)
+	_ encoding.TextUnmarshaler = new(ptrTextMarshaler)
 )

--- a/update.go
+++ b/update.go
@@ -59,9 +59,13 @@ func (u *Update) Range(name string, value interface{}) *Update {
 }
 
 // Set changes path to the given value.
+// If value is an empty string or nil, path will be removed instead.
 // Paths that are reserved words are automatically escaped.
 // Use single quotes to escape complex values like 'User'.'Count'.
 func (u *Update) Set(path string, value interface{}) *Update {
+	if isNil(value) {
+		return u.Remove(path)
+	}
 	path, err := u.escape(path)
 	u.setError(err)
 	expr, err := u.subExpr("🝕 = ?", path, value)

--- a/update_test.go
+++ b/update_test.go
@@ -73,3 +73,49 @@ func TestUpdate(t *testing.T) {
 		t.Error("expected ConditionalCheckFailedException, not", err)
 	}
 }
+
+func TestUpdateNil(t *testing.T) {
+	if testDB == nil {
+		t.Skip(offlineSkipMsg)
+	}
+	table := testDB.Table(testTable)
+
+	// first, add an item to make sure there is at least one
+	item := widget{
+		UserID: 4242,
+		Time:   time.Now().UTC(),
+		Msg:    "delete me",
+		Meta: map[string]string{
+			"abc": "123",
+		},
+		Count: 100,
+	}
+	err := table.Put(item).Run()
+	if err != nil {
+		t.Error("unexpected error:", err)
+		t.FailNow()
+	}
+
+	// update Msg with 'nil', which should delete it
+	var result widget
+	err = table.Update("UserID", item.UserID).Range("Time", item.Time).
+		Set("Msg", (*textMarshaler)(nil)).
+		Set("Meta.'abc'", nil).
+		Set("Meta.'ok'", (*ptrTextMarshaler)(nil)).
+		Set("Count", (*int)(nil)).
+		Value(&result)
+	if err != nil {
+		t.Error("unexpected error:", err)
+	}
+	expected := widget{
+		UserID: item.UserID,
+		Time:   item.Time,
+		Msg:    "",
+		Meta: map[string]string{
+			"ok": "null",
+		},
+	}
+	if !reflect.DeepEqual(result, expected) {
+		t.Errorf("bad result. %+v ≠ %+v", result, expected)
+	}
+}

